### DISCUSSION TITLE
M6 and M600

### DISCUSSION THREAD

**Discussion by coredump** [TooTall18T/tool_length_probe#12] [General]
If I understand correctly, using this configuration makes necessary to use `M600` and not `M6` to change tools, or am I missing something on the config?

**Comment by TooTall18T** [comment]
Yes that's right. You need to use M600 in your g-code.
M6 is still in use inside the subroutine.

**Comment by coredump** [comment]
Ah, ok, so if `M6` is still used inside the sub, there's on way to just remap `M6` to do the whole thing. Fair. I will find a way to rewrite the Fusion 360 post to use `M600`. Since it doesn't seem to be a parameter on the remap, I am guessing `M600` will use whatever you selected with `T#` previously.

**Reply by BSTguitars** [reply]
I was able to edit a post. I copied the Linuxcnc post , opened it and just found where it mentioned m6 and change it to m600.

I also made a webui that will open a g code file and let you replace text code with other code , highlighting it  also showing the new file for peace of mind  in the same window.

**Comment by TooTall18T** [comment]
The only way I know to change the action on M6 is to change the code of LinuxCNC.
I don't use Fusion so I can't help with that. But I think there is a way to change the postprocessor.

In your g-code you need to use "M600 T#" instead of "M6 T#".
M600 starts the subroutine and T# let the subroutine know wich tool you want to use next.

**Comment by coredump** [comment]
There are M6 remaps available: https://linuxcnc.org/docs/html/remap/remap.html#_remapping_tool_change_related_codes_t_m6_m61

And stuff that does it like https://forum.linuxcnc.org/show-your-stuff/41321-m6-tool-change-remapping-with-tool-length-measurement

**Comment by TooTall18T** [comment]
Oh thanks. It's good to know there is a other way than change the code of LinuxCNC.
But I think that is a part for a later version.

**Comment by TheRealBeef** [comment]
So I have both your intended m600 remap and also m6 remapped to the tool_touch_off.ngc, and it "appears" to have been working correctly for me in all cases. I may be misunderstanding some element of the logic and so there could be an edge case that I'm not properly accounting for.

In my ini:

REMAP= M6 modalgroup=6 ngc=tool_touch_off
REMAP= M600 modalgroup=6 ngc=m600

**Comment by TooTall18T** [comment]
Today I was able to check the remap M6 again. 

Yes the remap do the job and you are able to use M6 in you g-code to change the tools. 
I'm not sure how it works, because inside the subroutine is a second M6 and normally this would start a endless loob. But it doesn't happen.

But if you set a tool in the input window ([T 0] [M6 G43]) the machine moves directly over the probe sensor and waits for the tool change. If you stop that, some strange things happen. For example if there is no loaded g-code it opened a subroutine an I cannot close it or can see a loaded g-code.
The [M6 G43] button starts the subroutine and shows the message "Same tool" but the machine doesn't end the subroutine and you need to stop it by hand.

I don't know where Probe Basic still uses the normal M6 . You can try it but my recommendation is to use M600 and do not remap the M6 .

**Comment by coredump** [comment]
We where actually discussing that on the PrintNC discord, and someone pointed [this part of the documentation ](https://linuxcnc.org/docs/html/remap/remap.html#_making_minimal_changes_to_the_built_in_codes_including_code_m6_code), mainly:

> Since this might be a common scenario, the built in behavior of remapped codes has been made available within the remap procedure. The interpreter detects that you are referring to a remapped code within the procedure which is supposed to redefine its behavior. In this case, the built in behavior is used - this currently is enabled for the set: M6, M61,T, S, F). Note that otherwise referring to a code within its own remap procedure would be a error - a remapping recursion.

I noticed some of the behaviours you describe too, specially on some instances the routine seems to keep 'running' and I have to press `stop cycle` to be able to do anything else. I may go back to use M600 and just change the postprocessor.

**Comment by TheRealBeef** [comment]
> But if you set a tool in the input window ([T 0] [M6 G43]) the machine moves directly over the probe sensor and waits for the tool change. If you stop that, some strange things happen. For example if there is no loaded g-code it opened a subroutine an I cannot close it or can see a loaded g-code. The [M6 G43] button starts the subroutine and shows the message "Same tool" but the machine doesn't end the subroutine and you need to stop it by hand.

Ah this has happened to me. I guess I will change back to M600 only and see what happens.

**Comment by TooTall18T** [comment]
> We where actually discussing that on the PrintNC discord, and someone pointed [this part of the documentation ](https://linuxcnc.org/docs/html/remap/remap.html#_making_minimal_changes_to_the_built_in_codes_including_code_m6_code), mainly:
> 
Ahh, that make sence. It is good to know that. 

> I noticed some of the behaviours you describe too, specially on some instances the routine seems to keep 'running' and I have to press `stop cycle` to be able to do anything else. I may go back to use M600 and just change the postprocessor.

I think that would be the best way. There would be solutions to these behaviours, but it need more testing and maybe there is not a solution for every single one.
